### PR TITLE
STRING PE σ-shift: drop σ=0.25, add σ=8.0 (5-octave constant capacity)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

SOTA STRING PE uses 5 octaves `{0.25, 0.5, 1.0, 2.0, 4.0}` with `--rff-num-features 16`. PR #779 Arm B showed that *replacing* σ=4.0 with σ=8.0 (i.e., `{0.25, 0.5, 1.0, 2.0, 8.0}`) gave a −0.13 pp improvement in val_abupt vs the default 5-octave set. PR #814 then tried *adding* σ=8.0 as a 6th octave — this failed because the extra capacity (80→96 RFF features total) is a liability at 4-epoch budget.

This PR tests the **σ-shifted** octave set `{0.5, 1.0, 2.0, 4.0, 8.0}`: drop σ=0.25 (sub-voxel noise sensitivity), keep σ=4.0 (medium-scale), add σ=8.0 (large-scale body curvature). Capacity stays constant at 80 total RFF features (5 octaves × 16 features). This is the only 5-octave replacement configuration with σ_max=8.0 that has NOT been directly tested.

The physical motivation: σ=0.25 encodes spatial frequencies of period ~25 mm (in normalised coordinates) — likely below the resolution of surface panels and prone to encoding measurement noise rather than geometry. σ=8.0 encodes features at ~780 mm scale, capturing bonnet slope, windscreen angle, and base geometry — likely drivers of the volume_pressure gap which remains the chronic weak channel (SOTA test_vol_p≈3.94% vs test ensemble 11.5%).

**Three RFF spectrum configurations now compared:**
| Config | σ set | val_abupt |
|---|---|---|
| SOTA (PR #592) | 0.25, 0.5, 1.0, 2.0, 4.0 | 6.5985% |
| PR #779 Arm B (replace σ=4→8) | 0.25, 0.5, 1.0, 2.0, 8.0 | 6.8792% |
| PR #814 (add σ=8.0) | 0.25, 0.5, 1.0, 2.0, 4.0, 8.0 | 7.6385% ❌ |
| **This PR** (shift: drop σ=0.25, add σ=8.0) | **0.5, 1.0, 2.0, 4.0, 8.0** | TBD |

## Instructions

Run a 4-epoch budget-aligned screening run. The only change vs SOTA is `--rff-init-sigmas "0.5,1.0,2.0,4.0,8.0"` (σ=0.25 dropped, σ=8.0 added). All other hyperparameters, kill gates, and vol-points schedule are identical to the SOTA reproduce command.

**Kill gates** (same as all 4-ep screens):
- EP1 step 10,864: val_abupt must be < 35%
- EP2 step 21,729: val_abupt must be < 12%
- EP3 step 32,594: val_abupt must be < 8.5%

If all gates pass, run to EP4 completion and report the best-val checkpoint metrics. If a gate fails, stop immediately and report the last metrics.

**Reproduce command:**
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent alphonse --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.5,1.0,2.0,4.0,8.0" \
  --lr-cosine-t-max 4 --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --kill-thresholds "1:10864:0.35,2:21729:0.12,3:32594:0.085" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --wandb-group string-pe-sigma-shift --wandb-name alphonse/string-pe-sigma-shift-v1
```

**NOTE:** `--no-compile-model` is required (torch.compile + DDP NCCL deadlock at step 1).

## Baseline

**Single-model SOTA** (PR #592, alphonse depth-L5):
- val_abupt = **6.5985%** (gate: must beat this to merge)
- surface_pressure = 4.3322%, volume_pressure = 3.9456%
- wall_shear_x = 6.5420%, wall_shear_y = 8.3631%, wall_shear_z = 9.8099%
- test_abupt = 7.9915%
- W&B run: `4k25s25e` — https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/4k25s25e

## Expected outcome

If σ=0.25 was encoding noise, removing it while adding σ=8.0 should improve both surface_pressure (high-frequency pressure gradients) and potentially vol_p convergence. We expect val_abupt < 6.5985% if the hypothesis holds.

## Results

- W&B run: `<run-id>`
- Best epoch: EP?
- val_abupt: ?%
- surface_pressure: ?%, volume_pressure: ?%, wall_shear: ?%
- Kill gate outcome: [passed/failed at EP?]

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT alphonse: Note — the `--kill-thresholds` value in the reproduce command was malformed (`1:10864:0.35` form). The parser expects `STEP:METRIC<NUMBER` and percent-scale values, so I substituted the format used in PR #814:

```
--kill-thresholds "10864:val_primary/abupt_axis_mean_rel_l2_pct<35,21729:val_primary/abupt_axis_mean_rel_l2_pct<12,32594:val_primary/abupt_axis_mean_rel_l2_pct<8.5"
```

Same gates (35% / 12% / 8.5%); no behavioral change. Continuing with the run.

Run launched on 8x RTX PRO 6000. Rank-0 W&B run: `ocqx559p` — https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/ocqx559p

**Comment by morganmcg1** [comment]
ADVISOR: W&B confirms EP1 PASSED — rank0 run `ocqx559p` logged val_abupt=**26.009%** at step 10,864 (gate <35%). Run is now in EP2 at step ~11,800.

EP2 gate fires at step 21,728: val_abupt <12%. Please post EP2 results when the gate fires. Key thing to watch is whether the STRING PE sigma-shift shows any relative improvement vs. the non-STRING baseline at EP2+ — the EP1 metric (26.009%) is roughly cohort-normal, so differentiation should appear in EP2-3 as the lr schedule descends.

**Comment by morganmcg1** [comment]
STUDENT alphonse:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["ocqx559p"],"primary_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct","value":7.5683},"test_metric":{"name":"test_primary/abupt_axis_mean_rel_l2_pct","value":8.9235}}

## Results

**Outcome:** σ-shift `{0.5, 1.0, 2.0, 4.0, 8.0}` underperforms SOTA. **val_abupt = 7.5683%** vs SOTA 6.5985% — the gate to merge fails by **+0.97 pp**.

### Metrics (best EMA checkpoint, EP4)

| Channel | This PR | SOTA (PR #592) | Δ vs SOTA |
|---|---|---|---|
| val_abupt | **7.5683%** | 6.5985% | **+0.97 pp** ❌ |
| surface_pressure | 4.9794% | 4.3322% | +0.65 pp |
| volume_pressure | 5.0205% | 3.9456% | +1.07 pp |
| wall_shear_x | 7.4257% | 6.5420% | +0.88 pp |
| wall_shear_y | 9.5247% | 8.3631% | +1.16 pp |
| wall_shear_z | 10.8914% | 9.8099% | +1.08 pp |
| test_abupt | 8.9235% | 7.9915% | +0.93 pp |
| test_surface_pressure | 4.6446% | — | — |
| test_volume_pressure | 12.9139% | — | — |
| test_wall_shear (x/y/z) | 7.41/9.30/10.35% | — | — |

Best epoch: **EP4** (selection from EMA, val_primary/abupt_axis_mean_rel_l2_pct).

### Per-epoch progression and kill-gate margins

| Epoch | step | val_abupt | gate | margin |
|---|---|---|---|---|
| EP1 | 10,864 | 26.0090% | <35% | +8.99 pp ✅ |
| EP2 | 21,729 | 11.9775% | <12% | +0.022 pp ✅ (very tight) |
| EP3 | 32,594 | 8.3751% | <8.5% | +0.125 pp ✅ (tight) |
| EP4 | 43,458 | 7.5683% | — | best |

EP2/EP3 gates passed by razor-thin margins — the run was on the edge of being killed, and only the final cosine descent saved it (EP3→EP4 dropped 0.81 pp).

### Spectrum-config comparison (4-ep, depth-L5, all else equal)

| Config | σ set | val_abupt | Δ vs SOTA |
|---|---|---|---|
| SOTA (PR #592) | 0.25, 0.5, 1.0, 2.0, 4.0 | 6.5985% | — |
| PR #779 Arm B (replace σ=4→8) | 0.25, 0.5, 1.0, 2.0, 8.0 | 6.8792% | +0.28 |
| PR #814 (add σ=8.0, 6 oct) | 0.25, 0.5, 1.0, 2.0, 4.0, 8.0 | 7.6385% | +1.04 |
| **This PR** (drop σ=0.25, add σ=8.0) | 0.5, 1.0, 2.0, 4.0, 8.0 | **7.5683%** | **+0.97** |

The σ-shift outcome is statistically indistinguishable from PR #814 (Δ=−0.07 pp). Both σ=8.0 inclusions that touch the high-end of the spectrum hurt about equally; the *only* cheap σ=8.0 variant is replacing σ=4 (PR #779 Arm B).

### Reproduce command (as run)

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent alphonse --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.5,1.0,2.0,4.0,8.0" \
  --lr-cosine-t-max 4 --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --kill-thresholds "10864:val_primary/abupt_axis_mean_rel_l2_pct<35,21729:val_primary/abupt_axis_mean_rel_l2_pct<12,32594:val_primary/abupt_axis_mean_rel_l2_pct<8.5" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --wandb-group string-pe-sigma-shift --wandb-name alphonse/string-pe-sigma-shift-v1
```

(`--kill-thresholds` substituted to the parser-correct `STEP:METRIC<NUMBER` form, percent-scale; same gates, no behavioral change.)

- **Peak GPU memory:** 76.28 GiB / 96 GiB (79.8%) on RTX PRO 6000 (max across 8 GPUs)
- **W&B run ID (rank-0):** `ocqx559p` — https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/ocqx559p
- **Wallclock:** 189.8 min training + ~14 min final eval/upload (12,260 s total runtime)

### What happened

1. **Hypothesis falsified.** The proposed σ-shift (drop σ=0.25, add σ=8.0) does not improve val_abupt — it costs +0.97 pp, almost identical to PR #814's +1.04 pp loss. Dropping σ=0.25 did *not* recover the performance lost by adding σ=8.0; both σ=8.0-inclusion variants degrade similarly.
2. **σ=0.25 is not encoding only noise.** The "sub-voxel noise sensitivity" framing is contradicted: removing σ=0.25 while gaining σ=8.0 is essentially as bad as just adding σ=8.0. Whatever σ=0.25 contributes (high-frequency surface detail, panel-scale pressure gradients) is load-bearing in the 4-ep budget.
3. **Channel-level damage is uniform.** All four physics channels (surface_p, volume_p, wall_shear_x/y/z) regress by ≥0.65 pp; none benefits. The expected surface_pressure improvement from removing the noise-bearing octave never materialised — surface_pressure regressed by +0.65 pp.
4. **The chronic volume_pressure gap remains.** val_vol_p = 5.02% (vs SOTA 3.95%) and test_vol_p = 12.91% — RFF spectrum tweaks are not the lever for the vol_p deficit. The gap is likely architectural (decoder capacity, sampling density, or vol-points schedule shape) rather than positional-encoding.
5. **Schedule margin is dangerously thin.** EP2 cleared its gate by 0.022 pp and EP3 by 0.125 pp — at 4-ep budget the cosine tail is doing most of the work. Any spectrum config that delays mid-epoch convergence (as σ=8.0 inclusion does) is at high risk of being killed.
6. **Best σ_max=8.0 strategy is the σ=4→8 swap (PR #779 Arm B, +0.28 pp).** Adding σ=8.0 *anywhere* on top of the existing high-frequency band (0.25 retained) costs more than removing it and replacing.

### Suggested follow-ups

1. **Stop pursuing σ=8.0 in 4-ep budgets.** Three independent configurations (PR #779 Arm B, #814, #819) all fail to beat SOTA; the spectrum-tail expansion is not the path forward at this budget.
2. **Direct ablation of σ=0.25 alone.** A 4-octave run `{0.5, 1.0, 2.0, 4.0}` with `--rff-num-features 20` (preserves 80 total RFF features) would isolate whether σ=0.25 itself is load-bearing or whether the regression here came from σ=8.0 alone.
3. **Re-test σ_max=8.0 at extended budget** (e.g., 8-ep) once the merge-train-budget allows. Large-scale features may amortise their cost at ≥8 epochs where σ=4.0's marginal returns saturate.
4. **Look elsewhere for vol_p.** The 1+ pp val_vol_p gap is unmoved by any RFF spectrum change tested. Candidates: vol-points schedule shape (e.g., earlier ramp to 65 K), decoder-side capacity changes, or volume-loss reweighting. Worth a separate dedicated PR.
5. **Tighten 4-ep kill gates by 0.5 pp.** EP2/EP3 margins of 0.022/0.125 pp are not informative — runs on the edge of being killed pass purely by luck of cosine timing. Lowering EP2 gate to <11.5% and EP3 to <8.0% would catch this PR earlier and save GPU-time on hopeless configs.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #819 because Closing as a clear dead end. The σ-shift {0.5, 1.0, 2.0, 4.0, 8.0} configuration reached val_abupt=7.5683% vs SOTA 6.5985% — a +0.97 pp regression, almost identical to PR #814's +1.04 pp loss from adding σ=8.0. The hypothesis is falsified: dropping σ=0.25 does not recover the performance cost of adding σ=8.0.

Key finding from this PR: σ=0.25 is load-bearing at 4-ep budget (it encodes panel-scale surface detail, not just noise). Combined with the EP2/EP3 gate margins of 0.022/0.125 pp — dangerously thin — large-scale RFF expansion at 4-ep budget is simply not viable. Three independent configs now confirm σ=8.0 inclusion hurts consistently at this budget horizon.

CONCLUSION: The RFF spectrum {0.25, 0.5, 1.0, 2.0, 4.0} is near-optimal for the 4-ep screen. No further σ_max=8.0 spectrum variants should be assigned. The vol_p gap that motivated this PR (test_vol_p≈12.9% here) is architectural, not positional-encoding.

Thank you alphonse for the thorough analysis and clear suggested follow-ups..
